src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java

```diff
@@ -29,46 +34,97 @@ public ConnectionFlowDriver(FlowConnectionProvider connectionProvider,
   @Override
   public void run() {
     try {
-      PacketStep first = flowSteps.peek();
-      int previousPort = (first == null || first.getPort() == null)
-          ? portArgument : first.getPort();
+      FlowStep first = flowSteps.peek();
+      int previousPort = (first == null || first.getPorts().isEmpty())
+          ? portArgument : first.getPorts().get(0);
       LOG.info("starting new flow on {}", previousPort);
-      while (!flowSteps.isEmpty()) {
-        PacketStep step = flowSteps.poll();
-        if (step.getPort() != null && step.getPort() != previousPort) {
-          LOG.info("changing to connections on port {}", step.getPort());
-          previousPort = step.getPort();
-        }
-        FlowConnection flowConnection = connectionProvider.get(previousPort);
-        step.process(flowConnection);
-      }
+      processSteps(flowSteps, previousPort, connectionProvider);
       LOG.info("flow completed!");
-    } catch (ConnectionClosedException e) {
-      LOG.info("Connection closed by client while waiting for client packet");
-      if (e.getDiscardedPacket().getBytes().length > 0) {
-        LOG.debug("Discarding client packet {}", e.getDiscardedPacket(), e);
+    } catch (IOException | InterruptedException | ExecutionException e) {
+      handleProcessStepExceptions(e);
+    } finally {
+      try {
+        closeFlowConnections();
+      } catch (IOException e) {
+        LOG.error("Problem while releasing sockets", e);
       }
-    } catch (IOException e) {
+      MDC.clear();
+    }
+  }
+
+  private static void processSteps(Queue<FlowStep> flowSteps, int previousPort,
+      FlowConnectionProvider connectionProvider)
+      throws ExecutionException, InterruptedException, IOException {
+    while (!flowSteps.isEmpty()) {
+      FlowStep step = flowSteps.poll();
+      if (step instanceof ParallelPacketStep) {
+        LOG.info("starting parallel execution");
+        processParallelSteps((ParallelPacketStep) step, previousPort, connectionProvider);
+        continue;
+      }
+      PacketStep packetStep = (PacketStep) step;
+      if (!packetStep.getPorts().isEmpty() && packetStep.getPorts().get(0) != previousPort) {
+        LOG.info("changing to connections on port {}", packetStep.getPorts().get(0));
+        previousPort = packetStep.getPorts().get(0);
+      }
+      FlowConnection flowConnection = connectionProvider.get(previousPort);
+      packetStep.process(flowConnection);
+    }
+  }
+
+  private static void processParallelSteps(ParallelPacketStep step, int previousPort,
+      FlowConnectionProvider connectionProvider) {
+    List<List<FlowStep>> parallelSteps = step.getParallelSteps();
```

Comment:
we could rename parallelSteps to forks/branches, might it be less confusing? Keep it or leave it.

Reply:
I'll take this comment. Waaay more descriptive! Thanks 💪